tests/components/azure_devops/test_config_flow.py

```diff
@@ -0,0 +1,108 @@
+
+from tests.test_util.aiohttp import AiohttpClientMocker
+
+FIXTURE_USER_INPUT = {CONF_ORG: "random", CONF_PROJECT: "project", CONF_PAT: "abc123"}
+
+
+async def test_show_form(hass: HomeAssistant) -> None:
+    """Test that the setup form is served."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+
+async def test_authorization_error(hass: HomeAssistant) -> None:
+    """Test we show user form on Azure DevOps authorization error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"], FIXTURE_USER_INPUT,
+    )
+
+    assert result2["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result2["step_id"] == "user"
+    assert result2["errors"] == {"base": "authorization_error"}
+
+
+async def test_connection_error(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test we show user form on Azure DevOps connection error."""
+    aioclient_mock.get(
+        f"https://dev.azure.com/{FIXTURE_USER_INPUT[CONF_ORG]}/_apis/projects",
+        status=403,
+    )
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    with patch(
+        "homeassistant.components.azure_devops.config_flow.DevOpsClient.authorize",
+        side_effect=aiohttp.ClientError,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"], FIXTURE_USER_INPUT,
+        )
+
+    assert result2["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result2["step_id"] == "user"
+    assert result2["errors"] == {"base": "connection_error"}
+
+
+async def test_full_flow_implementation(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test registering an integration and finishing flow works."""
+    aioclient_mock.get(
+        f"https://dev.azure.com/{FIXTURE_USER_INPUT[CONF_ORG]}/_apis/projects/{FIXTURE_USER_INPUT[CONF_PROJECT]}",
+        json={"id": "abcd-abcd-abcd-abcd", "name": "project"},
+        headers={"Content-Type": "application/json"},
+    )
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    with patch(
+        "homeassistant.components.azure_devops.config_flow.DevOpsClient.authorized",
+        return_value=True,
```

Comment:
We also need to patch `async_setup_entry` in the integration to avoid setting up the entry and doing more I/O.